carla/recourse_methods/catalog/face/library/face_method.py

```diff
@@ -0,0 +1,282 @@
+    chosen_indeces = np.random.choice(
+        np.arange(0, index).tolist()
+        + np.arange(index + 1, data.values.shape[0]).tolist(),
+        replace=False,
+        size=number_samples,
+    )
+    chosen_indeces = [
+        index
+    ] + chosen_indeces.tolist()  # make sure sample under consideration included
+    data = data.iloc[chosen_indeces]
+    data = data.sort_index()
+    return data
+
+
+def build_constraints(data, i, keys_immutable):
+    """
+
+    Parameters
+    ----------
+    data: pd.DataFrame
+    i : int
+        Position of immutable key
+    keys_immutable: list[str]
+        Immutable feature
+
+    Returns
+    -------
+    np.ndarray, np.ndarray
+    """
+    epsilon = 0.5  # avoids division by 0
+    immutable_constraint_matrix = np.outer(
+        data[keys_immutable[i]].values + epsilon,
+        data[keys_immutable[i]].values + epsilon,
+    )
+    immutable_constraint_matrix1 = immutable_constraint_matrix / ((1 + epsilon) ** 2)
+    immutable_constraint_matrix1 = ((immutable_constraint_matrix1 == 1) * 1).astype(
+        float
+    )
+    immutable_constraint_matrix2 = immutable_constraint_matrix / (epsilon ** 2)
+    immutable_constraint_matrix2 = ((immutable_constraint_matrix2 == 1) * 1).astype(
+        float
+    )
+    return immutable_constraint_matrix1, immutable_constraint_matrix2
+
+
+def find_cf(
+    candidate_counterfactuals_star,
+    data,
+    immutable_constraint_matrix1,
+    immutable_constraint_matrix2,
+    index,
+    n,
+    y_positive_indeces,
+    is_knn,
+):
+    """
+    Steps 1 to 3 of the FACE algorithm
+
+    Parameters
+    ----------
+    candidate_counterfactuals_star: list
+    data: pd.DataFrame
+    immutable_constraint_matrix1: np.ndarray
+    immutable_constraint_matrix2: np.ndarray
+    index: int
+    n: int
+    y_positive_indeces: int
+    is_knn: bool
+
+    Returns
+    -------
+    None
+    """
+    # STEP 1 -- BUILD NETWORK GRAPH
+    graph = build_graph(
+        data, immutable_constraint_matrix1, immutable_constraint_matrix2, is_knn, n
+    )
+    # STEP 2 -- APPLY SHORTEST PATH ALGORITHM  ## indeces=index (corresponds to x^F)
+    distances, min_distance = shortest_path(graph, index)
+    # STEP 3 -- FIND COUNTERFACTUALS
+    # minimum distance candidate counterfactuals
+    candidate_min_distances = [
+        min_distance,
+        min_distance + 1,
+        min_distance + 2,
+        min_distance + 3,
+    ]
+    min_distance_indeces = np.array([0])
+    for min_dist in candidate_min_distances:
+        min_distance_indeces = np.c_[
+            min_distance_indeces, np.array(np.where(distances == min_dist))
+        ]
+    min_distance_indeces = np.delete(min_distance_indeces, 0)
+    indeces_counterfactuals = np.intersect1d(
+        np.array(y_positive_indeces), np.array(min_distance_indeces)
+    )
+    for i in range(indeces_counterfactuals.shape[0]):
+        candidate_counterfactuals_star.append(data.values[indeces_counterfactuals[i]])
```

Comment:
As posted above, better design here would be to assemble a list that is only valid for the scope of the function and then return it so it can all be appended together outside